src/consumers/common.js

```diff
@@ -2,9 +2,21 @@ const Promise = require("bluebird");
 
 const { EXHAUSTED } = require("../sequence");
 
-
-const INCOMPLETE = Symbol();
-
+function next (renderer, iter, push) {
+  const max = renderer.batchSize;
+  const nextVal = renderer._next();
+
+  if (nextVal === EXHAUSTED) {
+    return EXHAUSTED;
+  } else if (nextVal instanceof Promise) {
+    return nextVal
+    .then(push)
+    .then(() => iter < max && next(renderer, iter + 1, push));
+  } else {
+    push(nextVal);
+    return Promise.resolve(nextVal);
+  }
+}
```

Comment:
recursively build a chain of promises resolving to values to render so we can wait on an async cache get.